### ISSUE TITLE
Crackling in reconstructed sounds

### ISSUE THREAD

**Issue by previtus** [Jonathan-LeRoux/lws#12]
Hello Jonathan Le Roux,
after reading your paper I wanted to use your implemented algorithm in one of my projects.
I encountered strange **"crackling" noises** in the reproduced audio.

Here is a sample on Soundcloud: https://soundcloud.com/previtus/reconstruction-lws-example-free-sample-macleod-monkeys
(short part from the original is followed by the reconstructed audio)

---

The code I used is mostly inspired by the sample on your github page (the reconstruction part), it follows this:

1. loading audio using the librosa library
2. reconstruction same as the example on https://github.com/Jonathan-LeRoux/lws/tree/master/python (except for using "music" as a preset)

> lws_processor = lws.lws(window_size, hop_size, L=lws_L, fftsize=fft_size, mode="music")
3. reconstruction using lws_processor.istft back to audio signal
4. saving the results using scipy.io.wavfile.write(...)

The code in full is here (it's a minimal size demo of it): https://gist.github.com/previtus/7afcb81baa40cc3c50390a39210d7d30
The input file is:
- downloaded from https://files.freemusicarchive.org/storage-freemusicarchive-org/music/no_curator/Kevin_MacLeod/Best_of_2014/Kevin_MacLeod_-_Monkeys_Spinning_Monkeys.mp3
- converted by:
> ffmpeg -i MacLeodMonkeys.mp3 MacLeodMonkeys.wav

**Finally ... to the question:**
I wanted to ask, if you have encountered similar behaviour (crackling noises) and would know how to fix it. Or is it possible that I have made some errors using your code?
PS: I also tried different settings for the values of _fft_size, window_size, hop_size, sample_rate and lws_L_ but didn't see any improvement with the crackling noises.
Separately I was also testing a griffin lim algortithm with librosa's stft and istft and didn't get these issues (however got differently sounding artifacts and in general slightly slower run speed).

Thank you for any input and help, and also for all your work on the code,
Vit Ruzicka

**Comment by previtus** [comment]
PS: the output from the algorithm I was getting is:

Abs(X):  0.92 dB
LWS   : 19.22 dB

_(and additionally debug outputs from inside lws_processor.stft and lws_processor.istft - just to check the values:
STFT Debug: fsize,fshift,awin,fftsize,perfectrec =  2048 512  [1.41421356 1.41421356 1.41421356 ... 1.41421356 1.41421356 1.41421356] None True
iSTFT Debug: fshift,swin,awin,fftsize,perfectrec =  512 [0.1767767 0.1767767 0.1767767 ... 0.1767767 0.1767767 0.1767767] None None True)_

**Comment by Jonathan-LeRoux** [comment]
Hi Vit,

That is odd indeed, thanks for pointing this out. I have just tried your code both with the latest version of LWS (1.2.5) and the previous one (1.2.4), and I get this issue with 1.2.5 but not with 1.2.4, so it looks like I have introduced a bug when I attempted to support STFTs with hop sizes that do not divide the window length :/
Let me try to dig deeper into this, and sorry for the inconvenience. In the meantime, please use 1.2.4.
Thanks,
Jonathan

**Comment by previtus** [comment]
Hi Jonathan,

thank you for looking into it, you are right, the version 1.2.4 works well.
I think I can do what I need with this version, so as far as I am concerned this issue is solved.

Thanks,
Vit

**Comment by Jonathan-LeRoux** [comment]
Hi Vit,

I found the bug, I had a wrong formula when initializing the analysis window, which made it rectangular:
https://github.com/Jonathan-LeRoux/lws/blob/c15db11526eb39b808108623f9d8eb57c9f20e9e/python/lws.pyx#L370 
This line should have been `awin = np.sqrt( awin * synthwin(awin,fshift) )`.
Even without the bug, using a rectangular analysis window leads to this crackling.

I just pushed release 1.2.6 fixing this bug, it is now live on pypi.

Thanks for pointing this out!
Jonathan
